### DISCUSSION TITLE
Imported target "pybind11::headers" includes non-existent path

### DISCUSSION THREAD

**Discussion by hakonhagland** [pybind/pybind11#5091] [General]
Hi, I am using pybind11 2.11.1 on Ubuntu 23.10. I have installed the package `pybind11-dev`:

```
$ dpkg -L pybind11-dev
/.
/usr
/usr/include
/usr/include/pybind11
/usr/include/pybind11/attr.h
/usr/include/pybind11/buffer_info.h
/usr/include/pybind11/cast.h
/usr/include/pybind11/chrono.h
/usr/include/pybind11/common.h
/usr/include/pybind11/complex.h
/usr/include/pybind11/detail
/usr/include/pybind11/detail/class.h
/usr/include/pybind11/detail/common.h
/usr/include/pybind11/detail/descr.h
/usr/include/pybind11/detail/init.h
/usr/include/pybind11/detail/internals.h
/usr/include/pybind11/detail/type_caster_base.h
/usr/include/pybind11/detail/typeid.h
/usr/include/pybind11/eigen
/usr/include/pybind11/eigen/common.h
/usr/include/pybind11/eigen/matrix.h
/usr/include/pybind11/eigen/tensor.h
/usr/include/pybind11/eigen.h
/usr/include/pybind11/embed.h
/usr/include/pybind11/eval.h
/usr/include/pybind11/functional.h
/usr/include/pybind11/gil.h
/usr/include/pybind11/iostream.h
/usr/include/pybind11/numpy.h
/usr/include/pybind11/operators.h
/usr/include/pybind11/options.h
/usr/include/pybind11/pybind11.h
/usr/include/pybind11/pytypes.h
/usr/include/pybind11/stl
/usr/include/pybind11/stl/filesystem.h
/usr/include/pybind11/stl.h
/usr/include/pybind11/stl_bind.h
/usr/include/pybind11/type_caster_pyobject_ptr.h
/usr/lib
/usr/lib/cmake
/usr/lib/cmake/pybind11
/usr/lib/cmake/pybind11/FindPythonLibsNew.cmake
/usr/lib/cmake/pybind11/pybind11Common.cmake
/usr/lib/cmake/pybind11/pybind11Config.cmake
/usr/lib/cmake/pybind11/pybind11ConfigVersion.cmake
/usr/lib/cmake/pybind11/pybind11NewTools.cmake
/usr/lib/cmake/pybind11/pybind11Targets.cmake
/usr/lib/cmake/pybind11/pybind11Tools.cmake
/usr/share
/usr/share/doc
/usr/share/doc/pybind11-dev
/usr/share/doc/pybind11-dev/changelog.Debian.gz
/usr/share/doc/pybind11-dev/copyright
/usr/share/pkgconfig
/usr/share/pkgconfig/pybind11.pc
```

Note that, even if the list above says `/usr/lib/cmake/pybind11/pybind11ConfigVersion.cmake` the file is installed in `/lib/cmake/pybind11/pybind11ConfigVersion.cmake` since `/lib` is a symbolic link to `/usr/lib` on Ubuntu 23.10.

When I in a CMakeLists.txt run the command:

```
find_package(pybind11 2.2 CONFIG)
```

the file `/lib/cmake/pybind11/pybind11Config.cmake` will be executed and line 206 in that file:

```
get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../../../" ABSOLUTE)
```
I believe this is the same as this line in the GitHub repository:

https://github.com/pybind/pybind11/blob/f33f6afb667b6b5c0da7dee98dc02f51b4cc0e96/tools/pybind11Config.cmake.in#L201

The problem is that the above `get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../../../" ABSOLUTE)` assumes that `pybind11Config.cmake` is located in `/usr/lib/cmake/pybind11/pybind11Config.cmake` and not in ` /lib/cmake/pybind11/pybind11Config.cmake`. So PACKAGE_PREFIX_DIR will be set to "/" instead of "/usr" which will produce errors of the following type:

```
CMake Error in CMakeLists.txt:
  Imported target "pybind11::headers" includes non-existent path

    "/include"
```

**Comment by wjwwood** [comment]
I've run into this as well on Ubuntu 24.04, but strangely my colleagues didn't have this issue in a container and our CI doesn't run into this issue either (also using docker), so I'm not entirely sure why, but there seems to be another factor involved to trigger this issue, even though your explanation makes sense. I have the same kind of symlink on my machine, but so does my colleague and our CI:

```
$ ls -ls /lib
0 lrwxrwxrwx 1 root 7 Mar 30 17:04 /lib -> usr/lib
```

And here are my `pybind11-dev` and `cmake` versions:

```
$ dpkg -s pybind11-dev | grep Version
Version: 2.11.1-2
```

```
$ dpkg -s cmake | grep Version
Version: 3.28.3-1build7
```

**Comment by hakonhagland** [comment]
It seems to be related to the ordering of the `PATH` environment variable. I had the following PATH:

```
PATH=/bin:/usr/bin:/sbin:/usr/sbin
```

by moving the `/bin` entry from the beginning to the end like this:

```
export PATH=/usr/bin:/sbin:/usr/sbin:/bin
```

solved the problem for me.

**Comment by fpadula** [comment]
I know this is more than 1 year old, but I recently ran into this issue on Ubuntu 24.04 (ROS Jazzy container) and couldn't fix it using @hakonhagland's suggestion. I did manage to get it to compile by linking `/usr/include` to `/`, but this is not ideal IMO.

With the help of a coworker, we managed to get a working solution w/o any external modifications by adding the following line after `find_package(pybind11)`:

```CMake
set_target_properties(pybind11::pybind11_headers PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "/usr/include"
)
```

This essentially overwrites the offending `INTERFACE_INCLUDE_DIRECTORIES` property, making it compile w/o any issues.

**Comment by dkogan** [comment]
I just hit this too. It's a hard-to-debug thing, that wastes hours of time for those that hit it. Can somebody please take a look? Trivial reproducer. CMakeLists.txt:

```
cmake_minimum_required(VERSION 3.15)
project(test)

set(PYBIND11_FINDPYTHON ON)
find_package(pybind11 REQUIRED)
message(STATUS "pybind11_INCLUDE_DIR: ${pybind11_INCLUDE_DIR}")

pybind11_add_module(pyxxx tst.cc)
```

Then:

```
mkdir build
cd build
touch ../tst.cc
cmake ..
```

I see:

```
...
-- pybind11_INCLUDE_DIR: //include
-- Configuring done (1.0s)
CMake Error in CMakeLists.txt:
  Imported target "pybind11::headers" includes non-existent path

    "/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.

```

Happens on Debian boxes with `usermerge` installed. I.e. where this is true:

```
$ ls -lhd /lib
lrwxrwxrwx 1 root root 7 Oct 13  2022 /lib -> usr/lib
```
